go-controller/pkg/clustermanager/routeadvertisements/controller.go

```diff
@@ -280,6 +280,183 @@ func (c *Controller) reconcile(name string) error {
 	return c.updateRAStatus(ra, hadUpdates, err)
 }
 
+func subnetsOverlap(subnet1, subnet2 string) (bool, error) {
+	ipNets, err := util.ParseIPNets([]string{subnet1, subnet2})
+	if err != nil {
+		return false, err
+	}
+
+	// Two subnets overlap if one contains the other's IP or vice versa
+	return ipNets[0].Contains(ipNets[1].IP) || ipNets[1].Contains(ipNets[0].IP), nil
+}
+
+// checkSubnetOverlaps validates that subnets from the current RA don't overlap
+// with any existing subnets in the same VRFs, both within the RA itself and
+// across different RAs
+func (c *Controller) checkSubnetOverlaps(ra *ratypes.RouteAdvertisements, selectedNetworks *selectedNetworks) error {
+	vrfSubnets := createVRFSubnetsMap(ra, selectedNetworks)
+
+	// Check for overlaps within each VRF
+	for vrf, subnets := range vrfSubnets {
+		// Check for overlaps within current RA
+		if err := checkWithinRAOverlaps(subnets); err != nil {
+			return fmt.Errorf("%w: error checking subnet overlap on RouteAdvertisement %q in VRF %q: %w", errConfig, ra.Name, vrf, err)
+		}
+		// Check for overlaps between current RA and other RAs in the same VRF
+		if err := c.checkCrossRAOverlaps(ra, vrf, subnets); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// checkWithinRAOverlaps checks if subnets within the current RA overlap with each other
+func checkWithinRAOverlaps(subnets []string) error {
+	for i, subnet1 := range subnets {
+		for j, subnet2 := range subnets {
+			if i >= j {
+				continue
+			}
+			overlap, err := subnetsOverlap(subnet1, subnet2)
+			if err != nil {
+				return fmt.Errorf("failed to check subnet overlap between %s and %s: %w", subnet1, subnet2, err)
+			}
+			if overlap {
+				return fmt.Errorf("overlapping CIDR detected: %s and %s", subnet1, subnet2)
+			}
+		}
+	}
+	return nil
+}
+
+// createVRFSubnetsMap creates a map of VRF -> subnets from selectedNetworks
+func createVRFSubnetsMap(ra *ratypes.RouteAdvertisements, selectedNetworks *selectedNetworks) map[string][]string {
+	vrfSubnets := make(map[string][]string)
+	if ra.Spec.TargetVRF == "auto" {
+		for _, vrf := range selectedNetworks.vrfs {
+			network := selectedNetworks.networkVRFs[vrf]
+			subnets := selectedNetworks.networkSubnets[network]
+			vrfSubnets[network] = subnets
+		}
+	} else {
+		vrfSubnets[ra.Spec.TargetVRF] = selectedNetworks.subnets
+	}
+	return vrfSubnets
+}
+
+// checkCrossRAOverlaps checks if subnets from current RA overlap with subnets from other RAs in the same VRF
+func (c *Controller) checkCrossRAOverlaps(ra *ratypes.RouteAdvertisements, targetVRF string, currentRASubnets []string) error {
```

Comment:
Since a RAs with targetVRF == "auto" can have multiple VRFs we should pass the "vrfSubnets" instead of just targetVRF so we don't iterate RAs per entry "vrfSubnets"